<a href="https://colab.research.google.com/github/mightyyandres/AIDA2-ML/blob/main/58034_LabNo_02_BreananAndFriends.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Topic02a : Prelim Problem Set I

## Case 1
Represent the following representations into its vectorized form using LaTeX.
> **Problem 1.a. System of Linear Equations**
$$
\left\{
    \begin{array}\\
        -y+z=\frac{1}{32}\\ 
        \frac{1}{2}x -2y=0 \\
        -x + \frac{3}{7}z=\frac{4}{5}
    \end{array}
\right. $$
> **Problem 1.b. Linear Combination**
$$  \cos{(\theta)}\hat{i} + \sin{(\theta)}\hat{j} - \csc{(2\theta)}\hat{k}$$
> **Problem 1.c. Scenario**
>
>A conference has 200 student attendees, 45 professionals, and has 15 members of the panel. There is a team of 40 people on the organizing committee. Represent the *percent* composition of each *attendee* type of the conference in matrix form.

Express your answers in LaTeX in the answer area.


#Solution
> **Problem 1.a. System of Linear Equations (Solution)**
$$
{
    \begin{array}\\
        -y+z=\frac{1}{32}\\ 
        -y=\frac{1}{32}-z\\ 
        \frac{y}{-1}=\frac{1/32}{-1}-\frac{z}{-1}\\
        y=-\frac{1-32z}{32}\\ 
    \end{array}}
.$$
> **$$Substitute: y=-\frac{1-32z}{32}\\
\frac{1}{2}x -2(-\frac{1-32z}{32})=0 \\
\frac{1}{2}x +(\frac{1-32z}{16})=0 \\
\frac{1}{2}x =(-\frac{1-32z}{16}) \\
2.\frac{1}{2}x =2.(-\frac{1-32z}{16}) \\
x =-\frac{1-32z}{8} \\
Substitute: x =-\frac{1-32z}{8}\\
$$**


In [ ]:
import numpy as np

> **Problem 1.a. System of Linear Equations**

In [ ]:
A = np.array([[-1, 1],[1/2, 2],[-1, 3/7]])
A

array([[-1.        ,  1.        ],
       [ 0.5       ,  2.        ],
       [-1.        ,  0.42857143]])

In [ ]:
B = np.array([1/32, 4/5])
B


array([0.03125, 0.8    ])

In [ ]:
# X = np.linalg.inv(A).dot(B)
np.dot(A,B)

array([0.76875   , 1.615625  , 0.31160714])

> **Problem 1.b. Linear Combination**

In [ ]:
theta = np.radians(30)
R = np.array([[np.cos(theta), np.sin(theta),np.arcsin(2*theta)]])
R

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: invalid value encountered in arcsin
  


array([[0.8660254, 0.5      ,       nan]])

> **Problem 1.c. Scenario**

In [ ]:
Conference = np.array([200,45,15,40])
percentage = Conference/Conference.sum(axis=0)*100
print(percentage)

[66.66666667 15.          5.         13.33333333]


\begin{array}{ccc}
\text{Attendees}&\text{Number of People}&\text{Percentage}\\
Attendees&200&66.6\% \\
Professionals&45&15\% \\
Panels&15&5\%\\
Org\ Committee&40&13.33\%
\end{array}

# Case 2
> **Problem 2.a: Vector Magnitude**

>The magnitude of a vector is usually computed as:
$$||v|| = \sqrt{a_0^2 + a_1^2 + ... +a_n^2}$$
Whereas $v$ is any vector and $a_k$ are its elements wherein $k$ is the size of $v$.
Re-formulate $||v||$ as a function of an inner product. Further discuss this concept and provide your user-defined function.

> **Problem 2.b: Angle Between Vectors**

> Inner products can also be related to the Law of Cosines. The property suggests that:
$$u\cdot v = ||u||\cdot||v||\cos(\theta)$$
Whereas $u$ and $v$ are vectors that have the same sizes and $\theta$ is the angle between $u$ and $v$.

> Explain the behavior of the dot product when the two vectors are perpendicular and when they are parallel.


#Solution
> **Problem 2.a: Vector Magnitude**

In [ ]:
def vector_magnitude(x):
  return np.sqrt(sum(i**2 for i in x))

In [ ]:
a = 5*np.random.randn(6)
a

array([ 3.93964854, -7.15139717, -3.86628941, -1.8916867 , -0.48208711,
       -3.90560029])

In [ ]:
np.linalg.norm(a)

10.03374835526016

In [ ]:
vector_magnitude(a)

10.03374835526016

> **Problem 2.b: Angle Between Vectors**

In [ ]:
def angle_vectors(a,b):
  inner = np.inner(a, b)
  norms = np.linalg.norm(a) * np.linalg.norm(b)

  cos = inner / norms
  rad = np.arccos(np.clip(cos, -1.0, 1.0))
  deg = np.rad2deg(rad)
  print("Radiant: ", rad)
  print("Degree: ", deg)
  return deg,rad

In [ ]:
a = np.array([9, 5])
b = np.array([-6, 2])
print(a,b)

[9 5] [-6  2]


In [ ]:
angle_vectors(a,b)

Radiant:  2.3127435948008137
Degree:  132.51044707800082


(132.51044707800082, 2.3127435948008137)

# Case 3
For the final cases analysis we will be looking at series of equations building up a single feed-forward computation of a logistic regression. The case will not require you to learn fully what is logistic regression. 

$$X = \begin{bmatrix} 
— (x^{(1)})^T— \\ 
— (x^{(2)})^T— \\
\vdots \\
— (x^{(m)})^T— \\
\end{bmatrix} \text{, } 
Y = \begin{bmatrix} 
y^{(1)} \\ 
y^{(2)} \\
\vdots \\
y^{(m)} \\
\end{bmatrix} \text{, and } 
\theta = \begin{bmatrix} 
\theta^{(1)} \\ 
\theta^{(2)} \\
\vdots \\
\theta^{(m)} \\
\end{bmatrix} $$
The dataset $X$ has $m$ entries with $n$ features while $Y$ is the vector containing the groud truths of a the entries of $X$, and $\theta$ are the parameters or weights of the vectors. We first compute the vector product of the dataset and the parameters as:
$$ z = x^{(i)}\theta^{(i)} = X\cdot \theta\\_{\text{Eq. 3.1}}$$
We then solve for the hypothesis of the logistic regression alogrithm as:

$$ h_\theta(x) = g(z)\\_{\text{Eq. 3.2}}$$

Where $g$ is an acitvation function that maps the values of the hypothesis vector between a range of 0 and 1. We computed the activation as a sigmoid function:
$$g(z) = \frac{1}{1+e^{-z}}\\_{\text{Eq. 3.3}}$$
Finally we compute the loss of the logistic regression algorithm using $J$. Wheras $J(\theta)$ is a function that computes the logistic loss of the hypothesis with respect to the ground truths $y$. it is then computed as:
$$J(\theta) = \frac{1}{m} \sum^m_{i=0}=[-y^{(i)}\log({h_{\theta}(x^{(i)})})-(1-y^{(i)})\log(1-h_{\theta}(x^{(i)}))]\\_{\text{Eq. 3.4}}$$

> **Problem 3.a: Matrix Equivalences**

> In Eq. 1, $z$ can also be solved as $X \cdot \theta$ which is the vectorized form of $x^{(i)}\theta^{(i)}$. However, it can also be expressed as $\theta^T\cdot X$. Prove the equality of $X \cdot \theta$ with $\theta^T\cdot X$ in this case.

> **Problem 3.b: Matrix Shapes**

> Determine the shape of $h_\theta$ if $X$ has a shape of $(300,5)$.

> **Problem 3.c: Vectorization**

> Express $J(\theta)$ into its vectorized form.

> **Problem 4.c: Computational Programming (Also Laboratory 2)**

> Encode Equations 3.1 to 3.4 as the class `LRegression` wherein:

> * `LRegression` should be instantiated with a dataset $X$, a ground truth vector $y$, and a parameter vector $\theta$. Each parameter should have a data type of `numpy.array`.
> * It should further have `methods`reflecting to at least the four (4) aforementioned equations. Each should have a return value.


**Problem 3.a: Matrix Equivalences**

In [ ]:
import numpy as np

X = np.exp(50)
theta = np.exp(50)

z = np.dot(X,theta)

z

2.688117141816135e+43

In [ ]:

X = np.exp(50)
theta = np.exp(50)

z = np.dot(X,theta.T)

z

2.688117141816135e+43

$$X = \begin{bmatrix} 
— (x^{(1)})^T— \\ 
— (x^{(2)})^T— \\
\vdots \\
— (x^{(m)})^T— \\
\end{bmatrix}  \text{ and } 
\theta = \begin{bmatrix} 
\theta^{(1)} \\ 
\theta^{(2)} \\
\vdots \\
\theta^{(m)} \\
\end{bmatrix} $$

$$
\text{Computing for the matrix product X}\theta, \text{ we have:}
$$

$$X\theta = \begin{bmatrix} 
— (x^{(1)})^T\theta— \\ 
— (x^{(2)})^T\theta— \\
\vdots \\
— (x^{(m)})^T\theta— \\
\end{bmatrix}  \text{ = } 
\begin{bmatrix} 
— \theta^T(x^{(1)})— \\ 
— \theta^T(x^{(2)})— \\
\vdots \\
— \theta^T(x^{(m)})— \\
\end{bmatrix} $$
>$X \cdot \theta$ = $\theta^T\cdot X$
$$\begin{bmatrix} 
— (x^{(1)})^T— \\ 
— (x^{(2)})^T— \\
\vdots \\
— (x^{(m)})^T— \\
\end{bmatrix} . 
\begin{bmatrix} 
\theta^{(1)} \\ 
\theta^{(2)} \\
\vdots \\
\theta^{(m)} \\
\end{bmatrix} = 
\begin{bmatrix} 
\theta^{(1)} \ 
\theta^{(2)} \
\cdots \
\theta^{(m)} \
\end{bmatrix} \cdot \begin{bmatrix} 
— (x^{(1)})^T— \\ 
— (x^{(2)})^T— \\
\vdots \\
— (x^{(m)})^T— \\
\end{bmatrix}
$$
$\begin{bmatrix}(x^{(1)})^T \theta^{(1)} + (x^{(2)})^T \theta^{(2)} \cdots + (x^{(m)})^T \theta^{(m)}\end{bmatrix}$ = $\begin{bmatrix}(x^{(1)})^T \theta^{(1)} + (x^{(2)})^T \theta^{(2)} \cdots + (x^{(m)})^T \theta^{(m)}\end{bmatrix}$

**Problem 3.b**

$$ X = 
\begin{bmatrix} 
x^1 & x^2 & x^3 & x^4 & x^5 \\ 
x^6 & x^7 & x^8 & x^9 & x^{10} \\ 
\vdots & \vdots & \vdots & \vdots & \vdots \\ 
x^{296} & x^{297} & x^{298} & x^{299} & x^{300} 
\end{bmatrix} 
\text { , } 
\theta = 
\begin{bmatrix} 
(\theta^1) \\
(\theta^2) \\ 
\vdots \\ 
(\theta^k)
\end{bmatrix}
$$

$$
g(z) = \frac{1}{1+e^{-z}}
$$

$$
z =X \cdot \theta \\
h_{\theta} (x) = g(X\cdot\theta) \\
h_{\theta} (x) = g(z) \\
$$

$$
h_{\theta}(x) = 
\begin{bmatrix} 
\theta^{1}(x) & \theta^{2}(x) & \theta^{3}(x) & \theta^{4}(x) & \theta^{5}(x) \\ 
\theta^{6}(x) & \theta^{7}(x) & \theta^{8}(x) & \theta^{9}(x) & \theta^{10}(x) \\
\vdots & \vdots & \vdots & \vdots & \vdots \\ 
\theta^{296}(x) & \theta^{297}(x) & \theta^{298}(x) & \theta^{299}(x) & \theta^{300}(x) \\
\end{bmatrix} 
$$

**Problem 3.c**

$$J(\theta) = \frac{1}{m} \sum^m_{i=0}=[-y^{(i)}\log({h_{\theta}(x^{(i)})})-(1-y^{(i)})\log(1-h_{\theta}(x^{(i)}))]$$
$$\begin{bmatrix} 
\frac{\partial{J}}{\partial{\theta}_0} \\ 
\frac{\partial{J}}{\partial{\theta}_1} \\
\frac{\partial{J}}{\partial{\theta}_2} \\
\vdots \\
\frac{\partial{J}}{\partial{\theta}_n} \\
\end{bmatrix} = 
\frac{1}{m}
\begin{bmatrix} 
\sum^{m}_{i=0} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\
\sum^{m}_{i=0} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\sum^{m}_{i=0} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\vdots \\
\sum^{m}_{i=0} [ -y^{(i)} \log (h_{\theta} (x^{(i)})) - (1-y^{(i)}) \log (1-h_{\theta}(x^{(i)}))] \\ 
\end{bmatrix} 
$$

$$
\frac{\partial{J}(\theta)}{\partial{\theta}_0} =
\frac{1}{m}
\sum\limits_{i=1}^m (h_0(x^{(i)})-y^{(i)})x^{(i)}_j 
$$

 **Problem 4.c**

In [ ]:
import numpy as np

class LRegression:
  def _init_(self, dataset_X, truth_vector_y, vector_theta):
    self.dataset_X = dataset_X
    self.truth_vector_y = truth_vector_y
    self.vector_theta = vector_theta 
  
  def vector_magnitude(self):
    v = self.vector_theta 
    return np.sqrt(v@v)
  
  def vector_product(self):
    vec_prod = np.dot(self.dataset_X, self.vector_theta)
    return vec_prod

  def sigmoid_function(self):
    z = vector_product(self.dataset_X, self.vector_theta)
    g_z = (1 / (1 + np.exp(-z)))
    return  g_z

  def logistic_loss(self):
    # m = self.truth_vector_y.shape[0]
    # self.j_theta  = np.mean( np.dot(np.log(self.vector_theta), self.truth_vector_y.T) + np.dot(np.log(1- self.vector_theta), (1- self.truth_vector_y.T)))
    y = self.truth_vector_y.T
    h_x = sigmoid_function(self.vector_theta)
    j_theta  = np.mean((-y * np.log(h_x)) - (np.log(1- y) * (1- h_x)))
    return j_theta 

In [ ]:
X = np.random.random((300,5))*100
theta = np.random.random((300,1))
y = np.random.randint(0,2,(300,1))


In [ ]:
X

array([[52.02257548, 36.45326553, 27.22502003, 39.01985195, 22.85522448],
       [89.93327003, 14.20935048, 75.57121183, 22.98191749, 87.13371382],
       [34.46222551, 18.59028154, 86.91640386, 96.84439805, 69.23208048],
       ...,
       [ 1.44100726, 19.49218629,  5.12788156, 16.50077774,  9.19123307],
       [90.04867371, 86.73716211, 11.89180097, 47.73932295, 49.72988105],
       [76.60440174, 34.20085818, 73.42755315, 81.902814  , 71.58522157]])

In [ ]:
theta

array([[0.27581721],
       [0.12623991],
       [0.04468169],
       [0.25144714],
       [0.10819494],
       [0.14308798],
       [0.58088729],
       [0.3516727 ],
       [0.5606879 ],
       [0.11460707],
       [0.53777853],
       [0.78384459],
       [0.7537394 ],
       [0.94249598],
       [0.1290459 ],
       [0.27891158],
       [0.43040232],
       [0.98890349],
       [0.28879107],
       [0.97707139],
       [0.49777464],
       [0.86926332],
       [0.75766881],
       [0.56459052],
       [0.37217348],
       [0.31575377],
       [0.20541181],
       [0.59797902],
       [0.04220073],
       [0.55013063],
       [0.92981744],
       [0.34361319],
       [0.30407907],
       [0.78219666],
       [0.37275862],
       [0.35827462],
       [0.33788857],
       [0.49718934],
       [0.82109664],
       [0.9013687 ],
       [0.37980056],
       [0.76726557],
       [0.56006574],
       [0.30988171],
       [0.56393847],
       [0.92319373],
       [0.04045208],
       [0.693

In [ ]:
y

array([[1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [1],
       [0],
       [1],
       [0],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [0],
       [0],
       [1],
       [0],
       [1],
       [0],
    